File: server/src/Destiny/Roll.hs
module Destiny.Roll (rollAspect, rollStat) where

import Control.Monad.Random
import Data.List.Extra
import Destiny.Scene
import Destiny.World

import qualified Data.Map.Lazy as Map
import qualified Destiny.Timeline as Timeline

rollStat :: RandomGen r => RollId -> StatId -> World -> Rand r World
rollStat rid sid world@(World timeline events) = case Map.lookup sid stats of
    Just (Stat _ _ score) -> do
        -- TODO: Generate roll ID on the server and send it to just the client that initiated the
        -- roll.
        roll <- getRandomR (1, 6)
        return $ world { worldEvents = snoc events $ RollResult rid [roll, score] }
    Nothing -> return world
  where
    stats = sceneStats $ Timeline.value timeline

rollAspect :: RandomGen r => RollId -> AspectId -> World -> Rand r World
rollAspect rid aid world@(World timeline events) = case Map.lookup aid aspects of
    Just (Aspect { aspectDice = dice }) | dice >= 1 -> do
        roll <- getRandomR (1, 6)
        return world
            { worldTimeline = Timeline.modify timeline $ removeDie aid
            , worldEvents = updateEvents roll
            }
    _ -> return world
  where
    aspects = sceneAspects $ Timeline.value timeline
    updateEvents roll = case find requestedRoll events of
        Just _ -> map (amendRoll roll) events
        Nothing -> events
    amendRoll roll result@(RollResult rid' rolls)
        | rid == rid' = RollResult rid $ snoc rolls roll
        | otherwise   = result
    requestedRoll (RollResult rid' _)
        | rid == rid' = True
        | otherwise   = False

File: src/Server/Roll.fs
module internal Destiny.Server.Roll

open Destiny.Shared
open Destiny.Shared.Bag
open Destiny.Shared.Board
open Destiny.Shared.World
open System

let rollStat (random : Random) role statId rollId world =
    match Map.tryFind statId world.Board.Stats with
    | Some stat ->
        // TODO: Verify that the roll ID doesn't exist.
        let roll =
            { Id = rollId
              Role = role
              StatName = stat.Name
              StatResult = random.Next (1, 7)
              StatBase = stat.Score
              Invokes = [] }
        let rolls' =
            { Map = Map.add rollId roll world.Rolls.Map
              Order = List.add rollId world.Rolls.Order }
        { world with Rolls = rolls' }
    | None -> world

let rollAspect (random : Random) (Die role as die) aspectId rollId world =
    match Map.tryFind aspectId world.Board.Aspects with
    | Some aspect when Bag.contains die aspect.Dice ->
        let invoke =
            { Source = aspect.Description
              Role = role
              Result = random.Next (1, 7) }
        let addInvoke roll = { roll with Invokes = List.add invoke roll.Invokes }
        let rolls' = { world.Rolls with Map = Map.change addInvoke rollId world.Rolls.Map }
        { world with
              Board = Board.removeDie die aspectId world.Board
              Rolls = rolls' }
    | _ -> world
